crates/astria-sequencer/src/accounts/state_ext.rs

```diff
@@ -65,59 +69,131 @@ fn nonce_storage_key<T: AddressBytes>(address: T) -> String {
+        .context("failed to strip prefix from account balance key")?
+        .parse::<crate::storage_keys::hunks::Asset>()
+        .context("failed to parse storage key suffix as address hunk")?
+        .get())
+}
+
 #[async_trait]
 pub(crate) trait StateReadExt: StateRead + crate::assets::StateReadExt {
     #[instrument(skip_all)]
-    async fn get_account_balances(&self, address: Address) -> Result<Vec<AssetBalance>> {
+    fn account_asset_keys(
+        &self,
+        address: impl AddressBytes,
+    ) -> AccountAssetsStream<Self::PrefixKeysStream> {
         let prefix = format!("{}/balance/", StorageKey(&address));
-        let mut balances: Vec<AssetBalance> = Vec::new();
-
-        let mut stream = std::pin::pin!(self.prefix_keys(&prefix));
-        while let Some(Ok(key)) = stream.next().await {
-            let Some(value) = self
-                .get_raw(&key)
-                .await
-                .context("failed reading raw account balance from state")?
-            else {
-                // we shouldn't receive a key in the stream with no value,
-                // so this shouldn't happen
-                continue;
-            };
-
-            let asset = key
-                .strip_prefix(&prefix)
-                .context("failed to strip prefix from account balance key")?
-                .parse::<crate::storage_keys::hunks::Asset>()
-                .context("failed to parse storage key suffix as address hunk")?
-                .get();
-
-            let Balance(balance) =
-                Balance::try_from_slice(&value).context("invalid balance bytes")?;
-
-            let native_asset = self
-                .get_native_asset()
-                .await
-                .context("failed to read native asset from state")?;
-            if asset == native_asset.to_ibc_prefixed() {
-                balances.push(AssetBalance {
-                    denom: native_asset.into(),
-                    balance,
-                });
-                continue;
-            }
+        AccountAssetsStream {
+            underlying: self.prefix_keys(&prefix),
+        }
+    }
 
-            let denom = self
-                .map_ibc_to_trace_prefixed_asset(asset)
-                .await
-                .context("failed to get ibc asset denom")?
-                .context("asset denom not found when user has balance of it; this is a bug")?
-                .into();
-            balances.push(AssetBalance {
-                denom,
-                balance,
-            });
+    #[instrument(skip_all)]
+    fn account_asset_balances(
+        &self,
+        address: impl AddressBytes,
+    ) -> AccountAssetBalancesStream<Self::PrefixRawStream> {
+        let prefix = format!("{}/balance/", StorageKey(&address));
+        AccountAssetBalancesStream {
+            underlying: self.prefix_raw(&prefix),
         }
-        Ok(balances)
+    }
+
+    #[instrument(skip_all)]
+    async fn get_account_balances(
```

Comment:
This should probably not live in  `StateReadExt` since this is mempool-related logic.

I know you have the goal of keeping the concept of our state out of the mempol proper - I suggest you make a new sibling `mempool-state.rs`/`mod mempool_state` where you define a helper `fn get_account_balances<S: StateRead>(state: S, account: impl AddressBytes) -> HashMap<IbcPrefixed, u128>`.

This keeps both the mempool module and the `StateReadExt` impl clean from mempool requirements.

We should keep the `StateReadExt` implementations as thin wrappers around cnidarium. If a module (like mempool) needs extra structure on top of the stream, then that module should implement the logic there instead of here.